File: linked_list.rb
class LinkedList
  attr_accessor :head

  def initialize
    self.head = nil
  end

  # add to head
  def add(value)
    current_head = head
    self.head = Node.new(value, current_head)
  end

  # delete from head
  def delete(value)
    current = head
    previous = nil

    while current != nil
      if current.value == value
        if previous
          previous.next = current.next
        else
          self.head = current.next
        end
      end
      previous = current
      current = current.next
    end
  end

  def size
    count = 0
    current = head

    while current != nil
      current = current.next
      count += 1
    end

    count
  end

  def to_array
    array = []
    current = head
    while current != nil
      array << current.value
      current = current.next
    end
    array
  end
end

class Node
  attr_accessor :value, :next

  def initialize(value, next_node)
    self.value = value
    self.next = next_node
  end
end

File: data-structures/linked_list.py
class LinkedList:
    def __init__(self):
        self.head = None

    def add(self, value):
        self.head = Node(value, next_node=self.head)

    def delete(self, value):
        current = self.head
        previous = None
        found = False
        while current and not found:
            if current.value == value:
                found = True
            else:
                previous = current
                current = current.next

        if found:
            if previous != None:
                previous.next = current.next
            else:
                self.head = current.next

    def to_list(self):
        array = []
        current = self.head
        while current != None:
            array.append(current.value)
            current = current.next
        return array

    def size(self):
        count = 0
        current = self.head
        while current != None:
            count += 1
            current = current.next
        return count

class Node:
    def __init__(self, value, next_node=None):
        self.value = value
        self.next = next_node
